src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Azure.WebJobs.Host.Rpc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc
+{
+    /// <summary>
+    /// Endpoint data source which composes all the WebJobs extension data sources together.
+    /// </summary>
+    /// <remarks>
+    /// Implementation is adapted from <see cref="CompositeEndpointDataSource"/>.
+    /// https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/CompositeEndpointDataSource.cs.
+    /// </remarks>
+    internal sealed class ExtensionsCompositeEndpointDataSource : EndpointDataSource, IDisposable
+    {
+        private readonly object _lock = new();
+        private readonly List<EndpointDataSource> _dataSources = new();
```

Comment:
Is this a good candidate for lazy initialization?

Reply:
Theoretically we could, but I don't think there is much benefit for it.

1. This class is typically only ever 1 instance, so not much savings.
2. To get any savings, we can't use `Lazy<T>` (we would just be trading 1 object for another). We would need to remove `readonly` and then add more code to get back to the same level of nullability safety.

I don't think the work is worth it to save 1 list, which is not in a hot path.